src/LevelZeroShim.cpp

```diff
@@ -0,0 +1,632 @@
+        //  Right now that is only board_gpu subdevices
+        return m_num_board_gpu_subdevice;
+    }
+
+    int LevelZeroShimImp::frequency_domain_count(unsigned int device_idx, int domain) const
+    {
+        return m_devices.at(device_idx).subdevice.m_freq_domain.at(domain).size();
+    }
+
+    int LevelZeroShimImp::engine_domain_count(unsigned int device_idx, int domain) const
+    {
+        return m_devices.at(device_idx).subdevice.m_engine_domain.at(domain).size();
+    }
+
+    double LevelZeroShimImp::frequency_status(unsigned int device_idx, int domain, int domain_idx) const
+    {
+        return frequency_status_shim(device_idx, domain, domain_idx).actual;
+    }
+
+    LevelZeroShimImp::frequency_s LevelZeroShimImp::frequency_status_shim(unsigned int device_idx, int domain, int domain_idx) const
+    {
+        ze_result_t ze_result;
+        frequency_s result;
+
+        zes_freq_handle_t handle = m_devices.at(device_idx).subdevice.m_freq_domain.at(domain).at(domain_idx);
+        zes_freq_state_t state;
+        ze_result = zesFrequencyGetState(handle, &state);
+        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME, "LevelZeroShim::" + std::string(__func__) +
+                                                            ": Sysman failed to get frequency state", __LINE__);
+
+        result.voltage = state.currentVoltage;
+        result.request = state.request;
+        result.tdp = state.tdp;
+        result.efficient = state.efficient;
+        result.actual = state.actual;
+        result.throttle_reasons = state.throttleReasons;
+
+        return result;
+    }
+
+    double LevelZeroShimImp::frequency_min(unsigned int device_idx, int domain, int domain_idx) const
+    {
+        return frequency_min_max(device_idx, domain, domain_idx).first;
+    }
+
+    double LevelZeroShimImp::frequency_max(unsigned int device_idx, int domain, int domain_idx) const
+    {
+        return frequency_min_max(device_idx, domain, domain_idx).second;
+    }
+
+    std::pair<double, double> LevelZeroShimImp::frequency_min_max(unsigned int device_idx, int domain, int domain_idx) const
+    {
+        ze_result_t ze_result;
+        double result_min = 0;
+        double result_max = 0;
+
+        zes_freq_handle_t handle = m_devices.at(device_idx).subdevice.m_freq_domain.at(domain).at(domain_idx);
+        zes_freq_properties_t property;
+        ze_result = zesFrequencyGetProperties(handle, &property);
+        check_ze_result(ze_result, GEOPM_ERROR_RUNTIME, "LevelZeroShim::" + std::string(__func__) +
+                                                        ": Sysman failed to get domain properties.", __LINE__);
+        result_min += property.min;
+        result_max += property.max;
+
+        return {result_min, result_max};
+    }
+
+    uint64_t LevelZeroShimImp::active_time_timestamp(unsigned int device_idx, int domain, int domain_idx) const
+    {
+        return active_time_pair(device_idx, domain, domain_idx).second;
```

Comment:
Don't separate the pair to ensure these signals are accurate.  Pass the pair up, and the IOGroup will be responsible for interpreting.

This is a bigger deal for the batch interface.

Reply:
active_time_pair and energy_pair have been moved from private to public to provide this functionality to the devicepool and up to the IOGroup as needed.